### ISSUE TITLE
v0.1.0 illegal moves

### ISSUE THREAD

**Issue by tpoppins** [konsolas/ToppleChess#7]
Two examples from the ongoing 40/40 gauntlet, 58 games complete so far

```
[Event "Topple 0.1.0 64-bit 40/40"]
[Site "Dual E5-2690v2"]
[Date "2018.06.26"]
[Round "1"]
[White "Zeta Dva 0305 64-bit"]
[Black "Topple 0.1.0 64-bit"]
[Result "1-0"]
[ECO "B22"]
[GameDuration "00:59:13"]
[GameEndTime "2018-06-26T20:00:08.588 Eastern Daylight Time"]
[GameStartTime "2018-06-26T19:00:55.210 Eastern Daylight Time"]
[Opening "Sicilian"]
[PlyCount "121"]
[Termination "illegal move"]
[TimeControl "40/1200"]
[Variation "Alapin's Variation (2.c3)"]

1. e4 {book} c5 {book} 2. c3 {book} d6 {book} 3. d4 {book} Nf6 {book}
4. Bd3 {book} Nc6 {book} 5. Nf3 g5 {-0.68/14 32s} 6. Bxg5 {+1.07/12 28s}
Rg8 {-0.61/15 32s} 7. dxc5 {+1.27/11 33s} dxc5 {-0.39/15 32s}
8. Na3 {+1.28/11 33s} Bg4 {-0.31/14 32s} 9. h4 {+1.32/11 33s} h6 {-0.52/13 32s}
10. Bxf6 {+1.34/12 24s} exf6 {-0.43/15 32s} 11. Nc4 {+1.31/11 34s}
Qc7 {-0.51/15 32s} 12. Ne3 {+1.35/10 34s} O-O-O {-0.61/14 32s}
13. Qe2 {+1.36/11 34s} Be6 {-0.59/14 32s} 14. g3 {+1.52/11 34s}
Bd6 {-0.39/13 32s} 15. Bc4 {+1.50/11 22s} Bxg3 {-0.07/15 32s}
16. Bxe6+ {+2.62/12 33s} fxe6 {-0.03/15 32s} 17. fxg3 {+2.62/12 28s}
Qxg3+ {-0.18/15 32s} 18. Kf1 {+2.22/13 23s} Ne5 {+0.19/16 32s}
19. Nxe5 {+2.10/12 19s} fxe5 {+0.02/16 32s} 20. Ng2 {+1.77/12 27s}
Rd3 {+0.38/17 32s} 21. Kg1 {+1.75/12 36s} Rgd8 {+0.18/17 32s}
22. Rf1 {+1.59/12 36s} Rd2 {-0.12/16 32s} 23. Qf3 {+1.38/13 36s}
Qxf3 {-0.54/16 32s} 24. Rxf3 {+1.03/14 36s} Rxb2 {-0.49/17 31s}
25. a3 {+1.33/13 36s} Rdd2 {+0.18/16 33s} 26. Rf8+ {+1.26/13 28s}
Kd7 {0.00/17 33s} 27. Ne3 {+1.17/13 37s} Re2 {0.00/15 33s}
28. Rh3 {+1.35/14 36s} Ra2 {0.00/17 33s} 29. Rf7+ {+1.46/13 30s}
Kc6 {0.00/20 29s} 30. Rf6 {+1.43/13 37s} Ra1+ {0.00/20 33s}
31. Rf1 {+1.08/15 31s} Raa2 {0.00/21 33s} 32. Rhf3 {+1.08/13 38s}
b5 {+0.34/18 30s} 33. R1f2 {+1.09/14 29s} Rxf2 {-0.53/20 32s}
34. Rxf2 {+1.03/16 38s} Rxa3 {-0.42/21 33s} 35. Ng4 {+1.03/14 40s}
Rxc3 {-0.30/18 33s} 36. Nxe5+ {+1.06/13 40s} Kc7 {0.00/19 33s}
37. Ra2 {+1.13/13 40s} Kb7 {0.00/20 33s} 38. Nf7 {+0.97/13 40s}
b4 {+0.78/17 31s} 39. Nd8+ {+0.83/13 25s} Kb6 {+1.04/18 34s}
40. Kf2 {+0.63/14 28s} Rh3 {+0.99/20 33s} 41. Re2 {+0.63/15 29s}
Rxh4 {+1.12/17 30s} 42. Kg3 {+0.80/15 24s} Rh1 {+0.32/16 30s}
43. Nxe6 {+0.86/14 27s} b3 {+0.15/17 28s} 44. Nf4 {+0.88/13 31s}
Rc1 {+0.48/17 30s} 45. Rh2 {+0.98/12 31s} Kb5 {+0.21/17 30s}
46. e5 {+1.04/13 25s} Re1 {+0.61/18 30s} 47. e6 {+1.04/13 31s}
Kc6 {+0.72/17 30s} 48. Kg4 {+0.91/13 31s} a5 {+1.10/18 30s}
49. Rd2 {+0.69/13 30s} Re5 {+1.61/19 30s} 50. Rh2 {+0.52/13 31s}
h5+ {+1.89/19 30s} 51. Kf3 {+0.20/15 31s} c4 {+1.72/19 30s}
52. Rxh5 {+0.20/13 31s} Rxh5 {+5.58/23 30s} 53. Nxh5 {+0.10/14 31s}
Kd6 {+5.80/25 30s} 54. Nf6 {-7.90/14 31s} a4 {+6.26/26 30s}
55. Nd5 {-9.25/14 31s} a3 {+10.13/22 30s} 56. e7 {-12.00/13 31s}
Kd7 {+9.27/21 30s} 57. Ke3 {-12.30/13 22s} a2 {+11.41/21 29s}
58. Nb6+ {-12.30/12 32s} Kxe7 {+12.91/21 30s} 59. Nxc4 {-12.30/13 32s}
a1=Q {+15.33/21 30s} 60. Kd3 {-12.31/13 32s} Qf1+ {+M15/21 30s}
61. Kc3 {-12.31/13 32s, Black makes an illegal move: b3a2} 1-0

```
```
Event "Topple 0.1.0 64-bit 40/40"]
[Site "Dual E5-2690v2"]
[Date "2018.06.26"]
[Round "2"]
[White "Topple 0.1.0 64-bit"]
[Black "Clueless 1.4"]
[Result "0-1"]
[ECO "A05"]
[GameDuration "00:51:19"]
[GameEndTime "2018-06-26T20:23:34.999 Eastern Daylight Time"]
[GameStartTime "2018-06-26T19:32:15.302 Eastern Daylight Time"]
[Opening "Reti Opening"]
[PlyCount "106"]
[Termination "illegal move"]
[TimeControl "40/1200"]

1. Nf3 {book} Nf6 {book} 2. b3 {book} d5 {book} 3. Bb2 {book} e6 {book}
4. e3 {book} Bd6 {book} 5. d4 {+0.02/16 32s} O-O {+0.25/11 33s}
6. c4 {+0.03/16 30s} c5 {+0.73/10 33s} 7. Bd3 {+0.12/15 32s} dxc4 {+0.52/10 33s}
8. bxc4 {+0.12/15 32s} Nc6 {+0.41/10 33s} 9. O-O {+0.18/15 32s}
Qb6 {+0.34/10 33s} 10. Qe2 {+0.33/15 32s} Nb4 {+0.65/11 33s}
11. Rd1 {+0.45/15 32s} Rd8 {+0.78/10 33s} 12. dxc5 {+0.29/16 32s}
Bxc5 {+0.54/12 33s} 13. Bxf6 {+0.64/16 30s} gxf6 {+0.49/12 33s}
14. Nc3 {+0.62/16 33s} Be7 {+0.47/11 33s} 15. Rab1 {+0.84/14 33s}
Qa5 {+0.47/11 33s} 16. Rb2 {+0.78/13 33s} Rxd3 {+0.45/10 33s}
17. Rxd3 {+0.05/17 30s} Nxd3 {+0.42/11 33s} 18. Qxd3 {-0.09/15 33s}
b6 {+0.49/12 33s} 19. Rb5 {-0.03/14 33s} Qa3 {+0.54/13 33s}
20. Qc2 {-0.01/15 33s} Bb7 {+0.74/12 33s} 21. Rh5 {-0.33/14 33s}
f5 {+0.87/12 33s} 22. Nb5 {-0.56/14 33s} Qa5 {+0.94/11 33s}
23. Nbd4 {-0.59/14 33s} Rc8 {+1.19/11 33s} 24. g4 {-0.64/15 32s}
Bf6 {+1.51/10 33s} 25. gxf5 {-1.29/14 33s} Bxd4 {+3.77/11 33s}
26. Rg5+ {-1.17/17 33s} Kh8 {+3.31/12 33s} 27. fxe6 {-1.56/17 33s}
Qc3 {+3.29/12 33s} 28. Qxc3 {-1.46/19 33s} Bxc3 {+3.59/15 33s}
29. exf7 {-1.88/20 31s} Bxf3 {+3.68/14 33s} 30. Rf5 {-1.48/20 33s}
Kg7 {+3.88/15 33s} 31. Rxf3 {-1.80/20 33s} Kf8 {+3.92/15 33s}
32. Rf4 {-1.94/19 33s} Rc5 {+3.92/13 33s} 33. Kg2 {-1.69/18 33s}
Ra5 {+3.96/13 33s} 34. Rh4 {-1.80/18 33s} Rxa2 {+4.26/13 33s}
35. h3 {-1.96/18 33s} Be1 {+4.58/14 33s} 36. Rf4 {-2.05/19 33s}
a5 {+5.63/14 33s} 37. h4 {-2.52/18 33s} Rxf2+ {+11.45/14 33s}
38. Rxf2 {-12.99/25 33s} Bxf2 {+12.73/17 33s} 39. Kxf2 {-13.21/23 33s}
a4 {+13.10/18 33s} 40. Kf3 {-13.09/22 33s} a3 {+13.10/16 17s}
41. e4 {-13.41/21 30s} a2 {+13.36/15 30s} 42. Ke2 {-13.43/19 29s}
h5 {+14.53/16 49s} 43. Kd3 {-14.78/20 30s} a1=Q {+14.67/16 48s}
44. e5 {-14.75/20 30s} Qxe5 {+14.71/15 29s} 45. Kc2 {-14.97/20 30s}
Qf5+ {+19.47/15 29s} 46. Kc3 {-15.05/18 30s} Qf6+ {+23.39/16 29s}
47. Kd3 {-15.38/17 30s} Qxh4 {+24.78/16 29s} 48. Kc3 {-21.92/19 29s}
Qf6+ {+24.91/15 29s} 49. Kd3 {-24.19/19 30s} h4 {+M15/15 29s}
50. c5 {-M14/19 30s} h3 {+M13/14 29s} 51. cxb6 {-M12/27 30s} Qxb6 {+M11/14 3.4s}
52. Kc3 {-M10/56 29s} h2 {+M9/12 0.62s} 53. Kd3 {-M8/90 28s}
h1=Q {+M5/8 0.047s, White makes an illegal move: a2a4} 0-1

```

Topple-ivybridge.exe
Dual E5-2690v2 (IvyBridge, 20 cores) @3 GHz, 32 GB RAM, HT off
Win 7 x64 Pro
Cute Chess GUI 2018-06-25 x64 build
concurrency=18

Tirsa@CCRL

**Comment by konsolas** [comment]
I believe this is the issue:
 - board::parse_move determines whether a move is enpassant by checking if the target square is equal to the current enpassant square.
 - The default enpassant square (when no enpassant capture is possible) is 0, or A1
 - So board::parse_move interprets a promotion at A1 to be an enpassant move.
 - This means that, after the move is processed, the pawn is duplicated, leaving an extra pawn on the board, which results in illegal moves.

The fix will be just checking to ensure that the enpassant square is not zero before marking a move as being enpassant. 

Thank you very much for testing, I'll push a fix in a few minutes.

**Comment by konsolas** [comment]
Fixed in 55c6efd60af5a853242c3d31e401ceceeccf24bd v0.1.1

Thanks for the report
